src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs

```diff
@@ -533,22 +669,42 @@ internal static bool LocalFindAndPop(object callback)
                 return workItem;
             }
 
-            // Check for global work items
+            // Check for work items from the assigned global queue
+            if (s_assignableWorkItemQueueCount > 0 && tl.assignedGlobalWorkItemQueue.TryDequeue(out workItem))
+            {
+                return workItem;
+            }
+
+            // Check for work items from the global queue
             if (workItems.TryDequeue(out workItem))
             {
                 return workItem;
             }
 
+            // Check for work items in other assignable global queues
+            uint randomValue = tl.random.NextUInt32();
+            if (s_assignableWorkItemQueueCount > 0)
+            {
+                int queueIndex = tl.queueIndex;
+                for (int c = s_assignableWorkItemQueueCount, maxIndex = c - 1, i = (int)(randomValue % (uint)c);
+                    c > 0;
+                    i = i < maxIndex ? i + 1 : 0, c--)
+                {
+                    if (i != queueIndex && _assignableWorkItemQueues[i].TryDequeue(out workItem))
+                    {
+                        return workItem;
+                    }
+                }
+            }
+
             // Try to steal from other threads' local work items
             WorkStealingQueue localWsq = tl.workStealingQueue;
             WorkStealingQueue[] queues = WorkStealingQueueList.Queues;
-            int c = queues.Length;
-            Debug.Assert(c > 0, "There must at least be a queue for this thread.");
-            int maxIndex = c - 1;
-            uint i = tl.random.NextUInt32() % (uint)c;
-            while (c > 0)
+            Debug.Assert(queues.Length > 0, "There must at least be a queue for this thread.");
+            for (int c = queues.Length, maxIndex = c - 1, i = (int)(randomValue % (uint)c);
```

Comment:
A nit - I find the `for` with multiple assignments separated by commas difficult to read and the original way with the while seemed easier to follow to me.